gpkit/model.py

```diff
@@ -1,250 +1,424 @@
-    abstractable.
+    Example
+    -------
+    >>> gp = gpkit.Model(cost, constraints, substitutions)
 
 """
 
 import numpy as np
 
+from time import time
+from pprint import pformat
+from collections import defaultdict
 from functools import reduce as functools_reduce
-from operator import add
+from operator import mul, add
+from copy import deepcopy
 from collections import Iterable
 
-from .small_classes import Strings
-from .small_classes import PosyTuple
-from .small_classes import CootMatrix
-from .nomials import Constraint, Monomial
+from .nomials import Constraint, MonoEQConstraint
+from .nomials import Posynomial, Monomial, Signomial
 from .varkey import VarKey
+from .substitution import substitution, getsubs
 
-from .substitution import substitution
+from .small_classes import Strings
+from .small_scripts import flatten
 from .small_scripts import locate_vars
-from .small_scripts import is_sweepvar
 from .small_scripts import mag
+from .small_scripts import is_sweepvar
+from .small_scripts import sort_and_simplify
+
+from .solution_array import SolutionArray
+from .signomial_program import SignomialProgram
+from .geometric_program import GeometricProgram
+
+from .variables import Variable
+
+try:
+    from IPython.parallel import Client
+    CLIENT = Client(timeout=0.01)
+    assert len(CLIENT) > 0
+    POOL = CLIENT[:]
+    POOL.use_dill()
+    print("Using parallel execution of sweeps on %s clients" % len(CLIENT))
+except:
+    POOL = None
 
 
 class Model(object):
-    "Abstract class with substituion, loading / saving, and p_idx/A generation"
+    model_nums = defaultdict(int)
+
+    def __init__(self, cost=None, constraints=[],
+                 substitutions={}, *args, **kwargs):
+        if cost is None:
+            if not hasattr(self, "setup"):
+                raise TypeError("Models can only be created without a cost"
+                                " if they have a 'setup' method.")
+            try:
+                if "name" in kwargs:
+                    name = kwargs.pop("name")
+                else:
+                    name = self.__class__.__name__
+                setup = self.setup(*args, **kwargs)
+            except:
+                # TODO: pass error through!
+                print("The 'setup' method of this model had an error.")
+                raise TypeError
+            try:
+                cost, constraints = setup
+            except TypeError:
+                raise TypeError("Model 'setup' methods must return "
+                                "(cost, constraints).")
+        self.cost = cost
+        self.constraints = list(constraints)
+        self.substitutions = dict(substitutions)
+
+        if hasattr(self, "setup"):
+            # TODO: use super instead of Model?
+            k = Model.model_nums[name]
+            Model.model_nums[name] = k+1
+            name += str(k) if k else ""
+            for p in posynomials:
```

Comment:
Ha, yup, it's code from the old GP composability stuff, waiting for its own revamp.